src/pq_flash_index.cpp

```diff
@@ -1536,6 +1536,85 @@ template <typename T, typename LabelT> diskann::Metric PQFlashIndex<T, LabelT>::
     return this->metric;
 }
 
+template <typename T, typename LabelT> std::vector<T> PQFlashIndex<T, LabelT>::get_vector(std::uint64_t VID)
+{
+    ScratchStoreManager<SSDThreadData<T>> manager(this->thread_data);
+    auto this_thread_data = manager.scratch_space();
+    IOContext &ctx = this_thread_data->ctx;
+    std::vector<AlignedRead> read_reqs;
+    AlignedRead read;
+    char *buf = nullptr;
+    alloc_aligned((void **)&buf, SECTOR_LEN, SECTOR_LEN);
+    read.len = SECTOR_LEN;
+    read.buf = buf;
+    read.offset = NODE_SECTOR_NO(VID) * SECTOR_LEN;
+    read_reqs.push_back(read);
+    reader->read(read_reqs, ctx);
+
+#if defined(_WINDOWS) && defined(USE_BING_INFRA)
+    if ((*ctx.m_pRequestsStatus)[0] != IOContext::READ_SUCCESS)
+    {
+        aligned_free(buf);
+        return std::vector<T>();
+    }
+#endif
+
+    char *node_buf = OFFSET_TO_NODE(buf, VID);
+    T *node_coords = OFFSET_TO_NODE_COORDS(node_buf);
+
+    std::vector<T> out(get_data_dim());
+    memcpy(out.data(), node_coords, sizeof(T) * get_data_dim());
+
+    return out;
+}
+
+template <typename T, typename LabelT>
+std::vector<std::uint8_t> PQFlashIndex<T, LabelT>::get_pq_vector(std::uint64_t VID)
```

Comment:
Since the pq'ed vector is always in memory (without changing location), do you think we can just return a pointer to the underlying uint8_t array to save some efficiency?

Reply:
I don't think it's possible to make a vector in-place on an existing array without having very tricky rules about what happens if the vector is resized or the edited.

I replace it with call to the copy constructor in rvalue position, so it is guaranteed to be return-optimized